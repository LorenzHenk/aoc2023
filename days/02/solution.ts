type CubeCount = {
  count: number;
  color: 'red' | 'blue' | 'green';
};

type Draw = CubeCount[];

type Game = {
  id: number;
  draws: Draw[];
};

function parseCubeCount(input: string): CubeCount {
  // 3 blue
  const [countString, color] = input.trim().split(' ');

  return {
    color: color as CubeCount['color'],
    count: parseInt(countString),
  };
}

function parseDraw(input: string): Draw {
  // 3 blue, 4 red
  const cubeCountInputs = input.trim().split(',');

  return cubeCountInputs.map(parseCubeCount);
}

export function parseGame(input: string): Game {
  // Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
  const [gameDefinition, rest] = input.split(':');
  const gameId = parseInt(gameDefinition.split(' ').pop()!);

  const drawsInput = rest.split(';');

  return {
    id: gameId,
    draws: drawsInput.map(parseDraw),
  };
}

function parseInput(input: string): Game[] {
  const lines = input.split('\n');
  return lines.map(parseGame);
}

// only 12 red cubes, 13 green cubes, and 14 blue cubes?

const colorMaximumCountMap: Record<CubeCount['color'], number> = {
  red: 12,
  green: 13,
  blue: 14,
};

export function solvePart1(input: string): number {
  const parsedInput = parseInput(input);

  const validGames = parsedInput.filter(
    (value) =>
      !value.draws.some((draw) =>
        draw.some(
          (cubeCount) =>
            cubeCount.count > colorMaximumCountMap[cubeCount.color],
        ),
      ),
  );

  let sum = 0;
  for (let game of validGames) {
    sum += game.id;
  }

  return sum;
}

function getCubePower(game: Game): number {
  const cubeColorCounts: Record<CubeCount['color'], number> = {
    red: 0,
    blue: 0,
    green: 0,
  };

  for (let draw of game.draws) {
    for (let cubeCount of draw) {
      cubeColorCounts[cubeCount.color] = Math.max(
        cubeColorCounts[cubeCount.color],
        cubeCount.count,
      );
    }
  }

  return cubeColorCounts.blue * cubeColorCounts.green * cubeColorCounts.red;
}

export function solvePart2(input: string): number {
  const games = parseInput(input);

  let sum = 0;
  for (let game of games) {
    sum += getCubePower(game);
  }

  return sum;
}
