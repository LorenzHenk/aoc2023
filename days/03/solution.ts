type GearMapNumber = {
  value: number;
  xStart: number;
  xEnd: number;
  y: number;
};

type GearMapSymbol = {
  value: string;
  x: number;
  y: number;
};

type GearMap = {
  numbers: GearMapNumber[];
  symbols: GearMapSymbol[];
};

export function parseLine(line: string, y: number): GearMap {
  const result: GearMap = {
    numbers: [],
    symbols: [],
  };

  const trimmedLine = line.trim();

  let xStart: number | null = null;
  for (let index = 0; index < trimmedLine.length; index++) {
    const character = trimmedLine[index];

    if (/\d/.test(character)) {
      if (xStart === null) {
        xStart = index;
      }
    } else if (character === '.') {
      if (xStart !== null) {
        result.numbers.push({
          xStart: xStart,
          xEnd: index - 1,
          y: y,
          value: parseInt(trimmedLine.slice(xStart, index)),
        });

        xStart = null;
      }
    } else {
      if (xStart !== null) {
        result.numbers.push({
          xStart: xStart,
          xEnd: index - 1,
          y: y,
          value: parseInt(trimmedLine.slice(xStart, index)),
        });

        xStart = null;
      }

      result.symbols.push({
        value: character,
        x: index,
        y: y,
      });
    }
  }

  if (xStart !== null) {
    result.numbers.push({
      xStart: xStart,
      xEnd: trimmedLine.length - 1,
      y: y,
      value: parseInt(trimmedLine.slice(xStart, trimmedLine.length)),
    });
  }

  return result;
}

function parseInput(input: string): GearMap {
  const result: GearMap = {
    numbers: [],
    symbols: [],
  };

  input.split('\n').forEach((value, index) => {
    const lineResult = parseLine(value, index);
    result.numbers.push(...lineResult.numbers);
    result.symbols.push(...lineResult.symbols);
  });

  return result;
}

function findSymbolAtCoordinate(gearMap: GearMap, x: number, y: number) {
  return gearMap.symbols.find((symbol) => symbol.x === x && symbol.y === y);
}

function findSymbolsNextToNumber(
  gearMap: GearMap,
  gearMapNumber: GearMapNumber,
): GearMapSymbol[] {
  const symbols: GearMapSymbol[] = [];

  for (
    let index = gearMapNumber.xStart - 1;
    index <= gearMapNumber.xEnd + 1;
    index++
  ) {
    const symbol = findSymbolAtCoordinate(gearMap, index, gearMapNumber.y - 1);
    if (symbol) {
      symbols.push(symbol);
    }
  }

  const symbolLeftOfNumber = findSymbolAtCoordinate(
    gearMap,
    gearMapNumber.xStart - 1,
    gearMapNumber.y,
  );

  if (symbolLeftOfNumber) {
    symbols.push(symbolLeftOfNumber);
  }
  const symbolRightOfNumber = findSymbolAtCoordinate(
    gearMap,
    gearMapNumber.xEnd + 1,
    gearMapNumber.y,
  );

  if (symbolRightOfNumber) {
    symbols.push(symbolRightOfNumber);
  }

  for (
    let index = gearMapNumber.xStart - 1;
    index <= gearMapNumber.xEnd + 1;
    index++
  ) {
    const symbol = findSymbolAtCoordinate(gearMap, index, gearMapNumber.y + 1);
    if (symbol) {
      symbols.push(symbol);
    }
  }

  return symbols;
}

function filterGearMap(gearMap: GearMap): GearMapNumber[] {
  const result: GearMapNumber[] = [];

  for (let gearMapNumber of gearMap.numbers) {
    const symbolsNextToNumber = findSymbolsNextToNumber(gearMap, gearMapNumber);
    if (symbolsNextToNumber.length) {
      result.push(gearMapNumber);
    }
  }

  return result;
}

export function solvePart1(input: string): number {
  const parsedInput = parseInput(input);

  const filteredNumbers = filterGearMap(parsedInput);

  let sum = 0;
  for (let gearMapNumber of filteredNumbers) {
    sum += gearMapNumber.value;
  }

  return sum;
}

export function solvePart2(input: string): number {
  const gearMap = parseInput(input);

  gearMap.symbols = gearMap.symbols.filter((value) => value.value === '*');

  const symbolToNumberMap = new Map<GearMapSymbol, GearMapNumber[]>();

  for (let gearMapNumber of gearMap.numbers) {
    const symbols = findSymbolsNextToNumber(gearMap, gearMapNumber);

    for (let symbol of symbols) {
      const currentSymbolNumbers = symbolToNumberMap.get(symbol);
      if (currentSymbolNumbers === undefined) {
        symbolToNumberMap.set(symbol, [gearMapNumber]);
      } else {
        currentSymbolNumbers.push(gearMapNumber);
      }
    }
  }

  let sum = 0;
  for (let gearMapNumbers of symbolToNumberMap.values()) {
    if (gearMapNumbers.length === 2) {
      sum += gearMapNumbers[0].value * gearMapNumbers[1].value;
    }
  }

  return sum;
}
