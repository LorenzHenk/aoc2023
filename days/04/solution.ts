import { parseNumberString } from '../../util/parse';

interface Card {
  id: number;
  winningNumbers: number[];
  actualNumbers: number[];
}

export function parseLine(input: string): Card {
  // Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53

  const [, id, winningNumbersString, actualNumbersString] = input.match(
    /^Card +(\d+): ([\d ]+) +\| +([\d ]+)$/,
  ) as RegExpMatchArray;

  return {
    id: parseInt(id),
    winningNumbers: parseNumberString(winningNumbersString),
    actualNumbers: parseNumberString(actualNumbersString),
  };
}

export function parseInput(input: string) {
  const lines = input.split('\n');

  return lines.map(parseLine);
}

function getActualWinningNumbers(card: Card) {
  return card.actualNumbers.filter((num) => card.winningNumbers.includes(num));
}

export function solvePart1(input: string): number {
  const cards = parseInput(input);

  let sum = 0;

  for (let card of cards) {
    const actualWinningNumbers = getActualWinningNumbers(card);

    if (actualWinningNumbers.length) {
      sum += 2 ** (actualWinningNumbers.length - 1);
    }
  }

  return sum;
}

export function solvePart2(input: string): number {
  const cards = parseInput(input);

  const cardCopyCountByCardId: Record<number, number> = {};

  for (let card of cards) {
    const actualWinningNumbers = getActualWinningNumbers(card);

    const copyCount = cardCopyCountByCardId[card.id] ?? 0;

    for (let index = 1; index <= actualWinningNumbers.length; index++) {
      const nextCardIndex = card.id + index;
      const copyCountForNextCard = cardCopyCountByCardId[nextCardIndex] ?? 0;
      cardCopyCountByCardId[nextCardIndex] =
        copyCountForNextCard + (1 + copyCount);
    }
  }

  let totalCopyCount = 0;

  for (let cardCopyCount of Object.values(cardCopyCountByCardId)) {
    totalCopyCount += cardCopyCount;
  }

  return cards.length + totalCopyCount;
}
