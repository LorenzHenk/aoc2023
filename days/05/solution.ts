export type AlmanacMappingConfiguration = {
  destinationStartId: number;
  sourceStartId: number;
  rangeLength: number;
};

export type AlmanacMapping = {
  from: string;
  to: string;
  mappingConfigurations: AlmanacMappingConfiguration[];
};

export type Almanac = {
  seedIds: number[];
  mappings: AlmanacMapping[];
};

export function parseSeedBlock(input: string): number[] {
  // seeds: 79 14 55 13
  return input
    .trim()
    .split(': ')[1]
    .split(' ')
    .map((value) => parseInt(value));
}

export function parseMappingBlock(input: string): AlmanacMapping {
  // seed-to-soil map:
  // 50 98 2
  // 52 50 48

  const [firstLine, ...otherLines] = input.split('\n');

  const [, from, to] = firstLine.match(/(\w+)-to-(\w+)/)!;

  const mappingConfigurations: AlmanacMappingConfiguration[] = [];

  for (let line of otherLines) {
    // 50 98 2
    const [destinationStartId, sourceStartId, rangeLength] = line
      .trim()
      .split(' ')
      .map((value) => parseInt(value));

    mappingConfigurations.push({
      destinationStartId,
      sourceStartId,
      rangeLength,
    });
  }

  return { from, to, mappingConfigurations };
}

export function parseInput(input: string): Almanac {
  const blocks = input.split(/\n *\n/);
  const [seedBlock, ...mappingBlocks] = blocks;

  const seeds = parseSeedBlock(seedBlock);
  const mappings = mappingBlocks.map(parseMappingBlock);

  return { seedIds: seeds, mappings };
}

export function applyMapping(
  value: number,
  almanacMapping: AlmanacMapping,
): number {
  for (let mappingConfiguration of almanacMapping.mappingConfigurations) {
    if (
      value >= mappingConfiguration.sourceStartId &&
      value <
        mappingConfiguration.sourceStartId + mappingConfiguration.rangeLength
    ) {
      return (
        mappingConfiguration.destinationStartId +
        (value - mappingConfiguration.sourceStartId)
      );
    }
  }

  return value;
}

export function solvePart1(input: string): number {
  const almanac = parseInput(input);

  const locations = almanac.seedIds.map((seedId) => {
    // get location
    let currentValue = seedId;

    for (let mappingConfiguration of almanac.mappings) {
      currentValue = applyMapping(currentValue, mappingConfiguration);
    }

    return currentValue;
  });

  const minLocation = Math.min(...locations);

  return minLocation;
}

export function solvePart2(input: string): number {
  const almanac = parseInput(input);

  let minLocation = Infinity;

  for (let index = 0; index < almanac.seedIds.length; index += 2) {
    const baseSeedValue = almanac.seedIds[index];
    const seedRangeLength = almanac.seedIds[index + 1];
    for (let seedOffset = 0; seedOffset < seedRangeLength; seedOffset++) {
      const seedValue = baseSeedValue + seedOffset;

      // get location
      let currentValue = seedValue;

      for (let mappingConfiguration of almanac.mappings) {
        currentValue = applyMapping(currentValue, mappingConfiguration);
      }

      minLocation = Math.min(minLocation, currentValue);
    }
  }

  return minLocation;
}
