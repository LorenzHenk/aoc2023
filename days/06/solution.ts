import { parseNumberString } from '../../util/parse';

export type Race = {
  time: number;
  distance: number;
};

export type Game = {
  races: Race[];
};

export function parseInput(
  input: string,
  parseLine: (line: string) => number[],
): Game {
  // Time:      7  15   30
  // Distance:  9  40  200
  const [timeLine, distanceLine] = input.split(/\n/);

  const timeNumbers = parseLine(timeLine);
  const distanceNumbers = parseLine(distanceLine);

  const game: Game = {
    races: [],
  };

  if (timeNumbers.length !== distanceNumbers.length) {
    throw Error('Unexpected input');
  }

  for (let index = 0; index < timeNumbers.length; index++) {
    const time = timeNumbers[index];
    const distance = distanceNumbers[index];

    game.races.push({
      time,
      distance,
    });
  }

  return game;
}

function runRace(race: Race) {
  let winningDistances = [];
  for (let buttonHoldTime = 1; buttonHoldTime < race.time; buttonHoldTime++) {
    const driveTime = race.time - buttonHoldTime;

    const velocity = buttonHoldTime;
    const distance = velocity * driveTime;

    if (distance > race.distance) {
      winningDistances.push(distance);
    }
  }
  return winningDistances;
}

function getResult(game: Game) {
  let result = 1;

  for (let race of game.races) {
    const winningDistances = runRace(race);

    result *= winningDistances.length;
  }

  return result;
}

export function solvePart1(input: string): number {
  const game = parseInput(input, (line) =>
    parseNumberString(line.split(':')[1]),
  );

  return getResult(game);
}

export function solvePart2(input: string): number {
  const game = parseInput(input, (line) => [
    parseInt(line.split(':')[1].replaceAll(' ', '')),
  ]);

  return getResult(game);
}
