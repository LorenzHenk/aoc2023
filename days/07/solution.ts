const cardFaces = [
  'A',
  'K',
  'Q',
  'J',
  'T',
  '9',
  '8',
  '7',
  '6',
  '5',
  '4',
  '3',
  '2',
] as const;

export type Card = (typeof cardFaces)[number];

export type Hand = {
  cards: Card[];
  bid: number;
};

export type Game = {
  hands: Hand[];
};

export function parseLine(input: string): Hand {
  // 32T3K 765

  const [cardString, bidString] = input.trim().split(' ');

  return { bid: parseInt(bidString), cards: cardString.split('') as Card[] };
}

export function parseInput(input: string): Game {
  //   32T3K 765
  // T55J5 684
  // KK677 28
  // KTJJT 220
  // QQQJA 483

  const lines = input.split(/\n/);

  const game: Game = {
    hands: lines.map(parseLine),
  };

  return game;
}

const handsTypes = [
  'Five of a kind', // where all five cards have the same label: AAAAA
  'Four of a kind', // where four cards have the same label and one card has a different label: AA8AA
  'Full house', // where three cards have the same label, and the remaining two cards share a different label: 23332
  'Three of a kind', // where three cards have the same label, and the remaining two cards are each different from any other card in the hand: TTT98
  'Two pair', // where two cards share one label, two other cards share a second label, and the remaining card has a third label: 23432
  'One pair', // where two cards share one label, and the other three cards have a different label from the pair and each other: A23A4
  'High card', // where all cards' labels are distinct: 23456
] as const;

type HandsType = (typeof handsTypes)[number];

export function getHandValue(hand: Hand): HandsType {
  const counts = hand.cards.reduce(
    (acc, next) => {
      const existingCount = acc.find((value) => value.symbol === next);

      if (existingCount !== undefined) {
        existingCount.count += 1;
      } else {
        acc.push({ count: 1, symbol: next });
      }

      return acc;
    },
    [] as { count: number; symbol: Card }[],
  );

  if (counts.length === 1) {
    return 'Five of a kind';
  } else if (counts.some((countInfo) => countInfo.count === 4)) {
    return 'Four of a kind';
  } else if (
    counts.some((countInfo) => countInfo.count === 3) &&
    counts.some((countInfo) => countInfo.count === 2)
  ) {
    return 'Full house';
  } else if (counts.some((countInfo) => countInfo.count === 3)) {
    return 'Three of a kind';
  } else if (counts.filter((countInfo) => countInfo.count === 2).length === 2) {
    return 'Two pair';
  } else if (counts.some((countInfo) => countInfo.count === 2)) {
    return 'One pair';
  } else {
    return 'High card';
  }
}

export function solvePart1(input: string): number {
  const game = parseInput(input);

  const handsWithValues = game.hands.map((hand) => ({
    hand,
    value: getHandValue(hand),
  }));

  handsWithValues
    .sort((a, b) => {
      const aValue = handsTypes.indexOf(a.value);
      const bValue = handsTypes.indexOf(b.value);

      if (aValue === bValue) {
        for (let index = 0; index < 5; index++) {
          const aCard = a.hand.cards[index];
          const bCard = b.hand.cards[index];

          if (aCard !== bCard) {
            return cardFaces.indexOf(aCard) - cardFaces.indexOf(bCard);
          }
        }
      }

      return aValue - bValue;
    })
    .reverse();

  let result = 0;

  for (let index = 0; index < handsWithValues.length; index++) {
    result += handsWithValues[index].hand.bid * (index + 1);
  }

  return result;
}

const cardFacesWithJoker = [
  'A',
  'K',
  'Q',
  'T',
  '9',
  '8',
  '7',
  '6',
  '5',
  '4',
  '3',
  '2',
  'J',
] as const;

export function getHandValueWithJoker(hand: Hand): HandsType {
  const counts = hand.cards.reduce(
    (acc, next) => {
      const existingCount = acc.find((value) => value.symbol === next);

      if (existingCount !== undefined) {
        existingCount.count += 1;
      } else {
        acc.push({ count: 1, symbol: next });
      }

      return acc;
    },
    [] as { count: number; symbol: Card }[],
  );

  if (hand.cards.includes('J')) {
    const jokerCount = counts.find((countInfo) => countInfo.symbol === 'J')!;

    const maxCount = Math.max(
      ...counts
        .filter((countInfo) => countInfo.symbol !== 'J')
        .map((countInfo) => countInfo.count),
    );

    const maxCountCardInfo = counts.find(
      (countInfo) => countInfo.symbol !== 'J' && countInfo.count === maxCount,
    );

    if (maxCountCardInfo) {
      maxCountCardInfo.count += jokerCount.count;
      counts.splice(
        counts.findIndex((countInfo) => countInfo.symbol === 'J')!,
        1,
      );
    }
  }

  if (counts.some((countInfo) => countInfo.count === 5)) {
    return 'Five of a kind';
  } else if (counts.some((countInfo) => countInfo.count === 4)) {
    return 'Four of a kind';
  } else if (
    counts.some((countInfo) => countInfo.count === 3) &&
    counts.some((countInfo) => countInfo.count === 2)
  ) {
    return 'Full house';
  } else if (counts.some((countInfo) => countInfo.count === 3)) {
    return 'Three of a kind';
  } else if (counts.filter((countInfo) => countInfo.count === 2).length === 2) {
    return 'Two pair';
  } else if (counts.some((countInfo) => countInfo.count === 2)) {
    return 'One pair';
  } else {
    return 'High card';
  }
}

export function solvePart2(input: string): number {
  const game = parseInput(input);

  const handsWithValues = game.hands.map((hand) => ({
    hand,
    value: getHandValueWithJoker(hand),
  }));

  handsWithValues
    .sort((a, b) => {
      const aValue = handsTypes.indexOf(a.value);
      const bValue = handsTypes.indexOf(b.value);

      if (aValue === bValue) {
        for (let index = 0; index < 5; index++) {
          const aCard = a.hand.cards[index];
          const bCard = b.hand.cards[index];

          if (aCard !== bCard) {
            return (
              cardFacesWithJoker.indexOf(aCard) -
              cardFacesWithJoker.indexOf(bCard)
            );
          }
        }
      }

      return aValue - bValue;
    })
    .reverse();

  let result = 0;

  for (let index = 0; index < handsWithValues.length; index++) {
    result += handsWithValues[index].hand.bid * (index + 1);
  }

  return result;
}
