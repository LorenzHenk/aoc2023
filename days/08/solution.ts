type NodeName = string;

interface InstructionNode {
  nodeName: NodeName;
  left: NodeName;
  right: NodeName;
}

interface Instructions {
  navigation: Array<'R' | 'L'>;
  nodes: Record<NodeName, InstructionNode>;
}

function parseNodeLine(input: string): InstructionNode {
  // AAA = (BBB, CCC)
  // 0123456789012345
  const nodeName = input.slice(0, 3);
  const leftNodeName = input.slice(7, 10);
  const rightNodeName = input.slice(12, 15);

  return {
    nodeName,
    left: leftNodeName,
    right: rightNodeName,
  };
}

export function parseInput(input: string): Instructions {
  // RL
  //
  // AAA = (BBB, CCC)
  // BBB = (DDD, EEE)
  // CCC = (ZZZ, GGG)
  // DDD = (DDD, DDD)
  // EEE = (EEE, EEE)
  // GGG = (GGG, GGG)
  // ZZZ = (ZZZ, ZZZ)

  const lines = input.split(/\n/);

  const navigationInstructions = lines[0].split(
    '',
  ) as Instructions['navigation'];

  const nodes = lines.slice(2).map((line) => parseNodeLine(line.trim()));

  const nodesMap = nodes.reduce(
    (acc, next) => {
      return { ...acc, [next.nodeName]: next };
    },
    {} as Instructions['nodes'],
  );

  return { navigation: navigationInstructions, nodes: nodesMap };
}

export function solvePart1(input: string): number {
  const instructions = parseInput(input);

  let nextNode = instructions.nodes['AAA'];
  let steps = 0;

  while (nextNode.nodeName !== 'ZZZ') {
    const direction =
      instructions.navigation[steps % instructions.navigation.length];

    const nextNodeName = direction === 'L' ? nextNode.left : nextNode.right;

    nextNode = instructions.nodes[nextNodeName];
    steps++;
  }

  return steps;
}

export function solvePart2(input: string): number {
  const instructions = parseInput(input);

  const startNodes = Object.values(instructions.nodes).filter((node) =>
    node.nodeName.endsWith('A'),
  );

  let nextNodes = startNodes;
  let steps = 0;

  while (nextNodes.some((node) => !node.nodeName.endsWith('Z'))) {
    const direction =
      instructions.navigation[steps % instructions.navigation.length];

    const nextNodeNames = nextNodes.map((node) =>
      direction === 'L' ? node.left : node.right,
    );

    nextNodes = nextNodeNames.map((nodeName) => instructions.nodes[nodeName]);
    steps++;
    if (steps % 1000 === 0) {
      console.log(steps, nextNodes);
    }
  }

  return steps;
}
