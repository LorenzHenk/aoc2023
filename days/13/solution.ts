interface Field {
  rows: string[];
  columns: string[];
}

export function parseField(input: string): Field {
  const rows = input
    .trim()
    .split('\n')
    .map((line) => line.trim());

  const lineLength = rows[0].length;

  const columns = [];

  for (let x = 0; x < lineLength; x++) {
    columns.push(rows.map((row) => row[x]).join(''));
  }

  return { rows, columns };
}

export function parseInput(input: string): Field[] {
  const blockStrings = input.split(/\n\s*\n/);

  return blockStrings.map(parseField);
}

function findMirrorIndex(data: string[]): number | null {
  const totalLength = data.length;
  for (let index = 0; index < totalLength; index++) {
    const row = data[index];
    if (row === data[index + 1]) {
      let delta = 1;

      let isMirrored = true;

      while (index - delta >= 0 && index + 1 + delta < totalLength) {
        const left = data[index - delta];
        const right = data[index + 1 + delta];

        if (left !== right) {
          isMirrored = false;
          break;
        }

        delta++;
      }

      if (isMirrored) {
        return index + 1;
      }
    }
  }
  return null;
}

function compareLines(a: string, b: string): number {
  let differenceCount = 0;

  for (let index = 0; index < a.length; index++) {
    if (a[index] !== b[index]) {
      differenceCount++;
    }
  }

  return differenceCount;
}

function findMirrorIndexWithSmudge(data: string[]): number | null {
  const totalLength = data.length;
  for (let index = 0; index < totalLength - 1; index++) {
    const row = data[index];

    const initialLineErrorCount = compareLines(row, data[index + 1]);

    if (initialLineErrorCount > 1) {
      continue;
    }

    let errorCount = initialLineErrorCount;

    let delta = 1;

    let isMirrored = true;

    while (index - delta >= 0 && index + 1 + delta < totalLength) {
      const left = data[index - delta];
      const right = data[index + 1 + delta];

      errorCount += compareLines(left, right);

      if (errorCount > 1) {
        isMirrored = false;
        break;
      }

      delta++;
    }

    if (errorCount !== 1) {
      continue;
    }

    if (isMirrored) {
      return index + 1;
    }
  }
  return null;
}

function findFieldMirror(
  field: Field,
  solver: (data: string[]) => number | null,
  index: number,
): {
  direction: 'horizontal' | 'vertical';
  position: number;
} {
  const rowsMirror = solver(field.rows);
  const columnsMirror = solver(field.columns);

  if (rowsMirror !== null) {
    return { direction: 'horizontal', position: rowsMirror };
  }

  if (columnsMirror !== null) {
    return { direction: 'vertical', position: columnsMirror };
  }

  throw new Error("Couldn't find mirror at " + index);
}

export function solvePart1(input: string) {
  const fields = parseInput(input);

  const mirrorPositions = fields.map((field, index) =>
    findFieldMirror(field, findMirrorIndex, index),
  );

  let sum = 0;

  for (let mirrorPosition of mirrorPositions) {
    if (mirrorPosition.direction === 'horizontal') {
      sum += 100 * mirrorPosition.position;
    } else {
      sum += mirrorPosition.position;
    }
  }

  return sum;
}

export function solvePart2(input: string) {
  const fields = parseInput(input);

  const mirrorPositions = fields.map((field, index) =>
    findFieldMirror(field, findMirrorIndexWithSmudge, index),
  );

  let sum = 0;

  for (let mirrorPosition of mirrorPositions) {
    if (mirrorPosition.direction === 'horizontal') {
      sum += 100 * mirrorPosition.position;
    } else {
      sum += mirrorPosition.position;
    }
  }

  return sum;
}
