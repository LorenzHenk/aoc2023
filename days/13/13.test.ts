import { expect, describe, test } from '@jest/globals';
import { parseInput, solvePart1, solvePart2 } from './solution';

describe('13', () => {
  describe('part one', () => {
    test('input parsed', () => {
      const input = `#.##..##.
  ..#.##.#.
  ##......#
  ##......#
  ..#.##.#.
  ..##..##.
  #.#.##.#.

  #...##..#
  #....#..#
  ..##..###
  #####.##.
  #####.##.
  ..##..###
  #....#..#`;

      const result = parseInput(input);
      expect(result).toMatchSnapshot();
    });

    test('works with test input', () => {
      const input = `#.##..##.
      ..#.##.#.
      ##......#
      ##......#
      ..#.##.#.
      ..##..##.
      #.#.##.#.

      #...##..#
      #....#..#
      ..##..###
      #####.##.
      #####.##.
      ..##..###
      #....#..#`;

      const result = solvePart1(input);
      expect(result).toBe(405);
    });

    test('works with custom input', () => {
      const input = `#.##..##.
                    ..#.##.#.
                    ..#.##.#.
                    ##......#
                    ##......#`;

      const result = solvePart1(input);
      expect(result).toBe(400);
    });

    test('works with custom input', () => {
      const input = `..#.....##.
      .#..#..##.#
      .......###.
      ####.##.#.#
      ###.###....
      ###.###....
      ####.##.#.#
      .......####
      .#..#..##.#
      ..#.....##.
      ..#.....##.
      .#..#..##.#
      .......####`;

      const result = solvePart1(input);
      expect(result).toBe(1000);
    });
  });
  describe('part two', () => {
    test('works with test input', () => {
      const input = `#.##..##.
      ..#.##.#.
      ##......#
      ##......#
      ..#.##.#.
      ..##..##.
      #.#.##.#.

      #...##..#
      #....#..#
      ..##..###
      #####.##.
      #####.##.
      ..##..###
      #....#..#`;

      const result = solvePart2(input);
      expect(result).toBe(400);
    });
  });
});
