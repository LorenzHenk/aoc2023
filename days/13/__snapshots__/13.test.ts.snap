// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`13 part one input parsed 1`] = `
[
  {
    "columns": [
      "#.##..#",
      "..##...",
      "##..###",
      "#....#.",
      ".#..#.#",
      ".#..#.#",
      "#....#.",
      "##..###",
      "..##...",
    ],
    "rows": [
      "#.##..##.",
      "..#.##.#.",
      "##......#",
      "##......#",
      "..#.##.#.",
      "..##..##.",
      "#.#.##.#.",
    ],
  },
  {
    "columns": [
      "##.##.#",
      "...##..",
      "..####.",
      "..####.",
      "#..##..",
      "##....#",
      "..####.",
      "..####.",
      "###..##",
    ],
    "rows": [
      "#...##..#",
      "#....#..#",
      "..##..###",
      "#####.##.",
      "#####.##.",
      "..##..###",
      "#....#..#",
    ],
  },
]
`;
